### PR TITLE
[2.0.x] Enable hardware serial ports on LPC1768

### PR DESCRIPTION
**Pull request by tcm0116**
I got a new LPC1768-based control board that doesn't provide access to the native USB port on the LPC1768. As such, I have to use one of the UARTs for communication with the host. However, the HardwareSerial class didn't appear to work. I've re-written it based on the UART examples for the CMSIS framework and have done some additional cleanup to allow selection of the serial port.

Since `SERIAL_PORT` defaults to 0 in Configuration.h, which corresponds to UART0 on LPC1768 boards, I forced used of the USB Emulated Serial port by setting `SERIAL_PORT` to -1 in the _pins_RAMPS_RE_ARM.h_ file. I'm hoping to eventually add support for multiple host interfaces, at which time we can remove the hard-coded definition of `SERIAL_PORT` for Re-ARM boards.

UPDATE:

I'm one hour into a print on my mini Delta running this change, and it appears to be working as expected. The gcode commands are being sent to Marlin from a MKS TFT controller via a serial connection on UART0, with no apparent issues.

### CODE REVIEW

**Comment by tcm0116** [comment]
@Spawn32 if you checkout my branch for this PR (https://github.com/tcm0116/Marlin/tree/2.0.x-hardware_serial), your MKS TFT should be able to communicate with your MKS SBASE board via the UART. Keep in mind that you won't be able to send gcodes via the USB Emulated Serial Port, but you should still be able to access the SD Card from a PC via the USB.

**Comment by Spawn32** [comment]
will test it out tomorrow :)

**Comment by Spawn32** [comment]
Had to think for a second, i usually do a M80 to power on pc psu, TFT and motors from simplify32, but usb is dead now, i just forgot :)
But serial 0 seems dead, tested 250000 and 115000, MKS Tft doesn't even pickup the bed and hotend temp anymore..
Sd card was like normal...
My MKS tft can be setup in marlin, repiter and smoothie mode, dont know what and why there would be any different, tested in marlin mode now, can upload new setup for it and test smoothie mode..

**Comment by Spawn32** [comment]
Smoothie mode diden't help.
You are still using 
#define SERIAL_PORT 0
#define BAUDRATE 115200
in Configuration.h, or have you "hard coded" something for testing ? :)

**Comment by tcm0116** [comment]
For Re-ARM boards, it forces `SERIAL_PORT` to -1 in order to use the USB
port by default for now. For other boards, it uses the `SERIAL_PORT`
setting from _Configuration.h_. If `SERIAL_PORT` is -1, then it will echo
the output from Marlin on UART0, as it did before.

Can you try setting `SERIAL_PORT` to -1 to see if the MKS TFT can at least
see the temperatures?

**Comment by Spawn32** [comment]
usb works again with -1 but no temp reading on the display..

**Comment by tcm0116** [comment]
@Spawn32 that's odd. One possibility is that there's a pin conflict. The pins used for UART0 can also be used as ADC channels 6 and 7. Looking at your pins_MKS_SBASE.h file, it looks like you've only defined A0-A3. I haven't looked closely at how the HAL initializes the pins for the ADC channels, but there could be an error that is inadvertently initializing the pins for UART0. In my setup, I'm only initializing A0 or A1 for the extruder (I don't remember exactly). Are you using more than one extruder thermistor?

@p3p unless I'm mistaken, if someone were to enable the filament width sensor on Re-ARM, it would initialize P0.2 as an ADC channel, which would break UART0. Is that pin mapping really correct?

**Comment by tcm0116** [comment]
@Spawn32 it's also possible that I broke the debug framework since the MKS TFT should still see the temperature status when `SERIAL_PORT` is set to -1. Can you try using the current _frameworks/CMSIS/LPC1768/driver/debug_frmwrk.c_ from bugfix-2.0.x with my PR and `SERIAL_PORT` set to -1 to see if the MKS TFT is able to see the temperatures again?

**Comment by Spawn32** [comment]
Hmm... using the current debug_frmwrk.c broke something, sd card gone and my com port shows up for a few seconds before going over to unrecognized usb device..

Haven't re checked the ADC yet, but if you are sure it must be it i can have a look ?

**Comment by tcm0116** [comment]
What you described sounds like Marlin crashed and the watchdog timer reset
it to the bootloader. Not sure why it would do that. Did you have
SERIAL_PORT set to -1?

**Comment by Spawn32** [comment]
yup, set a -1

**Comment by tcm0116** [comment]
@Spawn32 I found an error with the ordering of the include files in the LPC1768 HAL main.cpp which was causing `MYSERIAL` to be defined differently in main.cpp than everywhere else. Please give this a try again and see if it works any better.

File: Marlin/src/HAL/HAL_LPC1768/HardwareSerial.h [hunk 1]
```diff
@@ -28,54 +28,41 @@
...
-#define LSR_RXFE  0x80
-
-#define UARTRXQUEUESIZE   0x10
-
 class HardwareSerial : public Stream {
 private:
-uint8_t PortNum;
-uint32_t baudrate;
+  static const size_t UARTRXQUEUESIZE = 256;
+
```
Comment (AnHardt) [inline Marlin/src/HAL/HAL_LPC1768/HardwareSerial.h]: https://github.com/MarlinFirmware/Marlin/blob/f6f881f5ed954bc2f44ca64895b6af99e74265a2/Marlin/Configuration_adv.h#L742-L746

Comment (AnHardt) [reply Marlin/src/HAL/HAL_LPC1768/HardwareSerial.h]: Having no TX-buffer nods also a comment in Configuration_adv.h

Comment (tcm0116) [reply Marlin/src/HAL/HAL_LPC1768/HardwareSerial.h]: I didn't realize those were defined in Configuration_adv.h. I'm working on getting them included and I'll also add support for buffered transmissions. I suppose we'll need to add XON/XOFF support at some point, but I'd like to get the basic serial implementation going first. 

From what I can tell, `RX_BUFFER_SIZE` can't actually be 0. Would you agree? If so, I'll update the comment in the Configuration_adv.h files as part of the change.

**Comment by Spawn32** [comment]
@tcm0116 when i tested yesterday i used my own cof.h and adv.h, have you done any mods to those or should i just test what you wrote yesterday ?

**Comment by tcm0116** [comment]
@Spawn32 you should be able to use your own configuration files. You'll
just need to pull the commit I pushed yesterday so you have the change I
made.

**Comment by Spawn32** [comment]
#define SERIAL_PORT 0
#define BAUDRATE 125000
No temp reading on TFT, no connection to simplify3d, sd-card ok

#define SERIAL_PORT 0
#define BAUDRATE 250000
No temp reading on TFT, no connection to simplify3d, sd-card ok

#define SERIAL_PORT -1
#define BAUDRATE 125000
Temp reading on TFT OK, but no fan control or movment, connection OK to simplify3d, sd-card ok

#define SERIAL_PORT -1
#define BAUDRATE 25000
Temp reading on TFT OK, but no fan control or movment, connection OK to simplify3d, sd-card ok

Actually i dont think it cares about baudrate when serial is -1, binary diden't update ...

But one thing i noticed was that the TFT dosent show the temp before i make a connection with simplify3d

**Comment by tcm0116** [comment]
>But one thing i noticed was that the TFT dosent show the temp before i make a connection with simplify3d

That makes sense because the debug framework only sends Marlin's output on the UART when `SERIAL_PORT` is -1. By default, Marlin only generates output in response to a command. Once you connect with Simplify3D, it starts requesting temperature status, which gets sent out over the USB and UART, so the TFT starts displaying the temps. 

I have no idea why the TFT can't communicate when you have `SERIAL_PORT` set to 0. It works on my control board and I know your UART can send data to the TFT since the debug framework is working. I'll keep looking. Could you send me your configuration files? Are you changing any other files when you load my branch?

**Comment by Spawn32** [comment]
i probably need to hook up a logger and se why it's actually sending ...
So we can't have host mode on usb and serial at the same time like smoothie seem's to have ?

**Comment by tcm0116** [comment]
>So we can't have host mode on usb and serial at the same time like smoothie seem's to have ?

Not currently. Marlin originally only supported a single serial port, and no one has worked on adding support for a second. 

One thing I noticed with the TFT display of that the first time the control board boots after copying a new firmware file to the SD card, it takes longer because it has to copy the file to flash, which causes the TFT display to timeout its connection and display the select baud rate screen. If you back out to the main screen, it then shows status. The next reboot works normally because it doesn't take as long to boot. Can you confirm that's not what going on?

**Comment by Spawn32** [comment]
fan 100% logged:
M106 S255 + line feed for marlin mode

Move X- logged:
G91 + line feed marlin mode
G1 X-1.0 F3000 + line feed marlin mode
G90 + line feed marlin mode

 Will set the TFT config for smoothie mode and se what the difference is.
The MKS TFT can only confgure it's baud by uploading a new config file to it's own sd-car..

**Comment by Spawn32** [comment]
Only difference i found in smoohie mode is that the display sends a : 
M114 + line feed when you enter the move menu....

EDIT: re tested marlin mode, it sends M114 when entering move menu as well, didn't find any difference..

One strange ting i found was first when i logged i was in 115200 8n1 none mode it was just showing me garbled data, set it to 8o1 none and it was ok, but going back to 8n1 none after that was still ok :)

**Comment by Spawn32** [comment]
omg, you are probably correct, this must have broken my serialport 0 :/
why on earth is re-arm using serial 0 as and adc ?

const adc_pin_data adc_pin_map[] = {
  {0, 23, 0},         //A0 (T0) - D67 - TEMP_0_PIN
  {0, 24, 1},         //A1 (T1) - D68 - TEMP_BED_PIN
  {0, 25, 2},         //A2 (T2) - D69 - TEMP_1_PIN
  {0, 26, 3},         //A3 - D63   MKS-SBASE FAN
  {1, 30, 4},         //A4 - D37 - BUZZER_PIN
  {1, 31, 5}          //A5 - D49 - SD_DETECT_PIN
  //{0,  3, 6},         //A6 - D0  - RXD0
  //{0,  2, 7}          //A7 - D1  - TXD0
};

but if this is the problem, doing what i did above will break the re-arm board want it ?

**Comment by tcm0116** [comment]
Did commenting those two lines fix it?

**Comment by Spawn32** [comment]
God, i am embarrassed now, and really feel i made a lot of extra work for you, work's 100% now...
How do we fix this in the pinmap_re_arm.h so we can used the same file for both boards ?

And last, who do i need to send flowers to so we can get dual host mode 🤣 

Again really sorry @tcm0116  :/

**Comment by p3p** [comment]
> why on earth is re-arm using serial 0 as and adc ?

@Spawn32 why on earth not? the pins are ADC capable so we mapped them as such. Although they shouldn't be enabled by default or interfere with the Hardware serial, (they were working fine for the debug hardware serial), I'm a bit out of the loop now with all the Arduino style classes that got merged.

**Comment by tcm0116** [comment]
Can you post the config files that weren't working? I want to add a sanity
check for conflicting pins.

**Comment by Spawn32** [comment]
you mean just the pinmap_re_arm.h files ?
thats the only one i changed, running serialport = 0 and baudrate 115200 in the other config..

@p3p dident mean it like that :) but i would not design a board that use rxd0/txt0 for something else, sbase, smoothie and duet all use the port as a serial port, probably the rest as well :)

**Comment by p3p** [comment]
@Spawn32 How does smoothieware handle multihost support? In theory it's impossible as you can not interlace two gcode streams and expect sane results, but If we don't care and assume that the user is never going to use both at the same time, we just have to merge to serial streams. 

edit: inside the LPC1768 platform would be simplest solution, well until you think about OKing only to the correct port.

**Comment by Spawn32** [comment]
@p3p haven't looked in to that, it's c++, not my strong side :/ can try to have  look, but both the display and simplify3d can be used at the same time...

**Comment by p3p** [comment]
what happens if you try and print from both would be an easy check ;), or just send move commands from the display while printing from the host ect,

**Comment by tcm0116** [comment]
@Spawn32 were you using the pinmap_re_arm.h file unmodified from my branch
when it wasn't working?

**Comment by Spawn32** [comment]
hahaha, could be fun, see your point :)
Will test that tomorrow and report back...

**Comment by Spawn32** [comment]
@tcm0116 yes, can't see you have made any changes to it after i committed it last week, so only change was the one i made to get it to work...

**Comment by tcm0116** [comment]
@Spawn32 I'm using the default pin mapping with my control board, and I
didn't have to comment out those two ADC channels. From what I can tell,
the ADC channels should only be initialized if you have a pin setup to use
that channel (such as the filament width sensor). Looking at the MKS SBASE
pins file you continued, I didn't see anything using ADC channels 6 or 7,
which are the conflicting channels.

I'd like to figure out what's different with your configuration so I can
add a sanity check to ensure that someone's not trying to use UART0 and ADC
channels 6 or 7 because of the fact that they share the same pins on the
LPC1768. If you have any time to help me figure out what feature or pin
setting is causing the ADC channels 6 or 7 to be initialized, that would be
very helpful.

**Comment by Spawn32** [comment]
Yes, of cores i will help you, it's 03.30 at night now, so have to be tomorrow, but here is the config files i used today, had a look and couldn't find anything special, filament is not used as i can se...
[Marlin.zip](https://github.com/MarlinFirmware/Marlin/files/1400526/Marlin.zip)
mks-sbase v1 2 pin layout
also added a picture og the used pins on the mKS SBASE...

**Comment by tcm0116** [comment]
No worries. I went ahead and put a preprocessor guard around those two ADC
channels in pinmap_re_arm.h if SERIAL_PORT is 0, which should at least
prevent them from being initialized as ADC channels. I'll check to see if
it throws any kind of error at compile time if one of those channels is
attempted to be used.

**Comment by tcm0116** [comment]
@Spawn32 The commit I just pushed adds the preprocessor guards to prevent use of ADC channels 6 and 7 when `SERIAL_PORT` is set to 0. If one of those channels is selected in the pins file, it won't throw an error at compile time, but does print an error message to the serial port during initialization indicating that the selected ADC channel is invalid.

@AnHardt The latest commit also supports the `RX_BUFFER_SIZE` and `TX_BUFFER_SIZE` settings provided in Configuration_adv.h. I tested various combinations of the settings, and it seems to be working well.

**Comment by luc-github** [comment]
Hi Just to confirm this PR enable the Serial 0 on my RE-ARM / Ramps 1.4 and Viki2

If I may suggest - I think RE_ARM_FORCE_SERIAL_PORT  add complexity and should be removed 
SERIAL_PORT is already defined why need to override it if another define is used ?
IMHO just explain -1 is SerialUSB for RE-ARM and may be others new board coming laters

```
/**
 * Select which serial port on the board will be used for communication with the host.
 * This allows the connection of wireless adapters (for instance) to non-default port pins.
 * Serial port 0 is always used by the Arduino bootloader regardless of this setting.
 *
 * :[0, 1, 2, 3, 4, 5, 6, 7] or -1 for Serial USB on RE-ARM
 */
#define SERIAL_PORT 0
```
this part : 
```
// For Re-ARM boards, always use the USB Emulated Serial Port unless RE_ARM_FORCE_SERIAL_PORT is defined
 +  #if MB(RAMPS_14_RE_ARM_EFB) || MB(RAMPS_14_RE_ARM_EEB) || MB(RAMPS_14_RE_ARM_EFF) || MB(RAMPS_14_RE_ARM_EEF) || MB(RAMPS_14_RE_ARM_SF)
 +    #ifndef RE_ARM_FORCE_SERIAL_PORT
 +      #undef SERIAL_PORT
 +      #define SERIAL_PORT -1
 +    #endif
 +  #endif
```
Is not necessary IMHO
But this is great PR - Great Jog 
thanks again

**Comment by tcm0116** [comment]
@luc-github The only reason I added `RE_ARM_FORCE_SERIAL_PORT` is because I didn't want to break everyone's Re-ARM setup as soon as this PR is merged. However, we could just remove that, add a comment in Configuration.h, and inform everyone in #7076 of the change. 

@p3p any thoughts on this?

**Comment by p3p** [comment]
I presume the majority of LPC1768 boards are going to need a example config so it won't be an issue for them, the Re-ARM being the exception because it doesn't really have a default config. but I think we can also assume that the majority of LPC1768 boards are going to use USB as the primary serial port .. would it be better to have Serial Port 0 mapped to the usb serial?

**Comment by tcm0116** [comment]
The boards themselves don't have example configurations, just printers. As
such, it could get confusing pretty quickly what port is what.

I was thinking we'd eventually get multi-host working, which would
alleviate the issue since Marlin could then support the USB emulated serial
and one of the UARTs.

Perhaps making using 0 for SERIAL_PORT to indicate the USB is a better
solution for now.

**Comment by luc-github** [comment]
get multi-host working would be definilty the best solution I think and very useful :+1:

**Comment by Nutz95** [comment]
@Spawn32 Hello! i've been struggling with this same matter for a few weeks now.
I saw through this thread that you managed to have your TFT display working with Re-arm and Ramps board which is what i'm looking for.

I have a Re-arm board and a MKS-tft screen.
I cannot have  them to communicate.
RIght now i've just performed a fresh pull from todays changes (21th of dec 2017) but no luck.
I also spent some time to look through the code for some clues.

My tft is connected to the AUX 1 pins of the Ramps 1.4 board.
The screen is set to 115200 bps
in marlin i have 
#define BAUDRATE 115200
#define SERIAL_PORT 0

-Right now i've tried any combination of Serial port index and Baudrate => not working
-i've also used a shorter cable => not working
-and i've also tried some voltage converters (3.3V<->5V) on the com pins => not working
-I've also tried to wire the screen on the side RX/TX ports of the Re-arm => not working
-I've also swapped the RX/TX pins just to check this part => not working

Any advice? tips on this? on my side i don't really need the USB serial, but i need the TFT screen instead.

**Comment by tcm0116** [comment]
@Nutz95

I'm fairly certain that the UART pins on the AUX1 port of the RAMPS board are not wired to the UART pins on the Re-ARM board. As such, you'll have to use the UART pins on the side of the Re-ARM board, which correspond to `SERIAL_PORT 0` in your configuration file. 

I currently have a USB-to-UART adapter connected to the UART pins on the side of my Re-ARM, and it works just fine. You shouldn't need any sort of adapter. 

Unfortunately, I can't seem to find any information on the pin out for the MKS TFT, so I can't really give you any further direction. Does the TFT have labels next to the connector for the ribbon cable that tell you what each pin does?

**Comment by Nutz95** [comment]
@Spawn32 Hello! i've been struggling with this same matter for a few weeks now.
I saw through this thread that you managed to have your TFT display working with Re-arm and Ramps board which is what i'm looking for.

I have a Re-arm board and a MKS-tft screen.
I cannot have  them to communicate.
RIght now i've just performed a fresh pull from todays changes (21th of dec 2017) but no luck.
I also spent some time to look through the code for some clues.

My tft is connected to the AUX 1 pins of the Ramps 1.4 board.
The screen is set to 115200 bps
in marlin i have 
#define BAUDRATE 115200
#define SERIAL_PORT 0

-Right now i've tried any combination of Serial port index and Baudrate => not working
-i've also used a shorter cable => not working
-and i've also tried some voltage converters (3.3V<->5V) on the com pins => not working
-I've also tried to wire the screen on the side RX/TX ports of the Re-arm => not working
-I've also swapped the RX/TX pins just to check this part => not working

Any advice? tips on this? on my side i don't really need the USB serial, but i need the TFT screen instead.

**Comment by Nutz95** [comment]
@tcm0116 yes this part is described on the PCB i can see the TX/RX/GND and +5V pin

**Comment by Spawn32** [comment]
@Nutz95 Sorry, i dont have a re-arm board, only a MKS-Sbase and my own STM32F7 board, but
test with what @tcm0116 told you and it should work.

If you have a serial to usb board you can connect that to the re-arm serial and open an terminal program to check whats being transmitted or not on the port..

**Comment by tcm0116** [comment]
@Nutz95 Here's the adapter I'm using:

KEDSUM® CP2102 Module STC Download Cable USB 2.0 to TTL 6PIN Serial
Converter For STC https://www.amazon.com/dp/B009T2ZR6W/ref=cm_sw_r_cp_
apap_wQEEdBDwkwKjt

**Comment by Nutz95** [comment]
@tcm0116, @Spawn32  I managed to have my re-arm communicating with my screen, but only when i have a serial connection attached to a pc.  So now i know the come is working a little bit, at least on 1 direction: from the Re-arm to the display.
I saw somewhere on some thread earlier where @Spawn32 managed to have his display working somehow.
I think i'm at the same state as you were on November but with a recent version of the code. SO i should have all the sanity check from @tcm0116 so far.
i tried to remove the use of ADC 6 and 7 from HAL pin mapping (by replacing 
#if SERIAL_PORT != 0
  #define P0_02 LPC1768_PIN(PORT(0), PIN( 2), INTERRUPT(1), PWM(0), ADC_CHAN(7))
  #define P0_03 LPC1768_PIN(PORT(0), PIN( 3), INTERRUPT(1), PWM(0), ADC_CHAN(6))
#endif
to this
#define P0_02 LPC1768_PIN(PORT(0), PIN( 2), INTERRUPT(1), PWM(0), ADC_NONE)
#define P0_03 LPC1768_PIN(PORT(0), PIN( 3), INTERRUPT(1), PWM(0), ADC_NONE)

inside file src\HAL\HAL_LPC1768/pinmapping.h

, but i don't see any effects. The display can't send any data to the re-arm and so it cannot read back board status.
Any tips on how you managed it to work? 
did you still have USB connection installed?

**Comment by tcm0116** [comment]
The `#if SERIAL_PORT != 0` above those two lines will prevent anything from being able to use the UART0 pins when you have `SERIAL_PORT` set to 0, which is exactly why they're there. You shouldn't need to change anything in the code. 

Marlin currently only allows you to use either USB or one of the UARTS, as configured by `SERIAL_PORT`. However, you can give my multi-material host branch (https://github.com/tcm0116/Marlin/tree/2.0.x-multi_host) a try if you want to see what happens when you use both USB and the UART. You might be able to get some clues about what's going on with the TFT.

**Comment by Nutz95** [comment]
@tcm0116 i've tried your branch and it's fixing my issues, now my tft display is communicating in both directions.
I'll install this on my printer and print directly from the display to see if there is any side effects on my setup.
Thanks again!

**Comment by Nutz95** [comment]
@tcm0116 i could test your branch on my printer, my display issues are fixed now.
I re-used my previous config files that i merged with your branch, and i see some behavior changes in my axis movements.

at leas t my X an Y axis are inverted for the same config file, and all my stepper motors have timing issues
On X and Y axis the motors seems to have infinite acceleration (my config is set to 1000).
on Z and E axis it seems the communication timing with my DRV8825 drivers is not good anymore (i've already set MINIMUM_STEPPER_PULSE to 2 in my config_adv.h file)

Do you know what could have changed in the firmware to explain this change in behavior?

**Comment by tcm0116** [comment]
@Nutz95 there's an issue with the current 2.0.x branch (which my multi-host branch is caught up with) that's causing issues with the motion. You can try adding the changes from #8878, which should resolve the motion issue.

**Comment by Nutz95** [comment]
@tcm0116 Thanks, i've applied your revert change and this fixes my issue!

**Pull request merged**